### PR TITLE
stock_background_job

### PR DESCRIPTION
**Pull request by Malak-abuthaa**

### CODE REVIEW

File: myapp/scheduler/scheduler.py
Comment (murradkh) [inline myapp/scheduler/scheduler.py:1]: you need to add the package name and the version into the `requirements.txt ` file, in order to be noticed and run by the team

File: myapp/scheduler/stock_api_update.py
Comment (murradkh) [inline myapp/scheduler/stock_api_update.py]: the code in this method is kind of the same as the code in method `update_top_stocks` ...
recommended to centralize the code in one place and avoid duplications

File: myapp/scheduler/stock_api_update.py
Comment (murradkh) [inline myapp/scheduler/stock_api_update.py]: showing many repetitive logs to the output, I think it's unneeded

File: myapp/apps.py
Comment (murradkh) [inline myapp/apps.py]: a better place to start the thread is in `commands/stock_manager.py`, to be run whenever the command "runserver" provided, like this you get:
1) a more logical place to search for the thread initialization.
2) no need for hard-coded condition (`if 'runserver' not in sys.argv`)
3) more reliability on starting the service.  

it's just a refactoring thing, you can keep it as it is if you want

**Review by murradkh — COMMENTED** [review]
a couple of changes needed i wrote on the code, and there are errors in the log when opening the web-page:

Screenshot 2020-10-14 144916

File: myapp/scheduler/scheduler.py
Comment (Malak-abuthaa) [reply myapp/scheduler/scheduler.py:1]: how did got these errors , did u start the server with --noreload ?

File: myapp/scheduler/stock_api_update.py
Comment (kamalq97) [inline myapp/scheduler/stock_api_update.py]: I might be better if you catch a specific exception and try to raise it instead of ignoring it. Maybe @murradkh could give a second opinion on this.

File: myapp/scheduler/scheduler.py
Comment (kamaql97) [reply myapp/scheduler/scheduler.py:1]: So now we can `run python manage.py` runserver without the extra `--noreload` argument?

**Review by kamalq97 — COMMENTED** [review]
I have run the server and I can verify that the scheduled updating is working on my machine. 

Make sure to update this branch from the master branch to enable us to do some manual integration testing and ensure nothing has been adversely affected by the changes in this pull request. 

There are also changes we could make on the front-end to highlight the fact the stocks data is being updated every 10 seconds, but that can be done in the future with a separate pull request (maybe I can work on it 🙂)

**Comment by murradkh** [comment]
> I have run the server and I can verify that the scheduled updating is working on my machine.
> 
> Make sure to update this branch from the master branch to enable us to do some manual integration testing and ensure nothing has been adversely affected by the changes in this pull request.
> 
> There are also changes we could make on the front-end to highlight the fact the stocks data is being updated every 10 seconds, but that can be done in the future with a separate pull request (maybe I can work on it 🙂)

@kamaql97 I think tommer said we need to do manual testing after it's merged it into the master, and the code owner needs to handle this.

File: myapp/management/commands/stock_manager.py
Comment (murradkh) [inline myapp/management/commands/stock_manager.py]: great!  now it's cleaner and centralized code

File: myapp/scheduler/stock_api_update.py
Comment (murradkh) [inline myapp/scheduler/stock_api_update.py:8]: here two exceptions could happen needs to handle (update your branch with the master to see how I handle exceptions from the service)

File: myapp/scheduler/stock_api_update.py
Comment (murradkh) [reply myapp/scheduler/stock_api_update.py]: in this block of code can happen different exceptions, the `db_lock` exception, and `key_value` exception (missing attributes sent by the service),  and maybe more.
I think that's okay for `db_lock` exception to ignore it because there are no actions we need to do.
for other exceptions, as kamal said, we need to know about it and see it in logs.

File: myapp/scheduler/stock_api_update.py
Comment (murradkh) [inline myapp/scheduler/stock_api_update.py:6]: I'm curious about this decorator is doing

**Comment by murradkh** [comment]
the code looks better, no errors in the logs, just a small improvement, to make the stocks data appears in the initial web-page loading without needing to reload the page, i.e. to add to the "home" view to run your code first, and then return index.html.

File: myapp/scheduler/stock_api_update.py
Comment (murradkh) [reply myapp/scheduler/stock_api_update.py:6]: I read about `transaction.atomic`,  it's a good feature for aborting transactions in case any exception occurs; in your case, I think it's unneeded, because you already catching the exception, thus the `transaction.atomic` would not work.
see this for more information: 
https://stackoverflow.com/questions/34730385/django-rollback-save-with-transaction-atomic

File: myapp/scheduler/stock_api_update.py
Comment (Malak-abuthaa) [reply myapp/scheduler/stock_api_update.py:8]: you mean exception when api called ? to check if the API called fine ?

Comment (murradkh) [reply myapp/scheduler/stock_api_update.py:8]: yes, see in `get_stock_info`  function in "`stock_api`" file  what exceptions might happen

Comment (Malak-abuthaa) [reply myapp/scheduler/stock_api_update.py:8]: @murradkh the two exception already defined ! , can you please specify what changes are required ? i'm confused

File: myapp/apps.py
Comment (murradkh) [inline myapp/apps.py:9]: getting an error when running  `makemigrations`

**Review by murradkh — COMMENTED** [review]
looks fine, just a small comment added
can you also add the logs for sending/receiving data from the service please, sorry that's i told to you remove, I realized it's a good indication of service status.

File: myapp/apps.py
Comment (Malak-abuthaa) [reply myapp/apps.py:9]: Hi Murrad , i run this command on my machine , i got no error , can you please tell me what the error exactly ?

File: myapp/scheduler/stock_api_update.py
Comment (kamalq97) [inline myapp/scheduler/stock_api_update.py:25]: It is good that you are now catching a specific exception. Is this the only error that could occur here?

File: myapp/apps.py
Comment (murradkh) [reply myapp/apps.py:9]: try to clear your DB, and then run `makemigrations` 
Screenshot 2020-10-18 125607
that's happen because the thread which fetching data from the service trying to insert them into not existed table "`myapp_stock`"

File: myapp/scheduler/scheduler.py
Comment (kamalq97) [inline myapp/scheduler/scheduler.py:8]: Why has the job been added twice?

**Review by kamalq97 — APPROVED** [review]
Good job! I have run the server and can confirm it is working without any apparent issues. I am not sure why the line `scheduler.add_job(stock_api_update)` has been added to the **scheduler.py** file though.

Aside from the above and the better error handling in **stock_api_update.py**, are there any other major changes since my last review? (I might have missed some of the modifications you made since a large of files has been changed since you updated this branch from the master repository).

**Review by murradkh — APPROVED** [review]

File: myapp/scheduler/scheduler.py
Comment (Malak-abuthaa) [reply myapp/scheduler/scheduler.py:8]: hi Kamal , i've add comments  on slack , i've explained why i did this

**Comment by murradkh** [comment]
@Malak-abuthaa Two things that you shouldn't do in any project, ignoring the test/check status, and not resolving conflicts appropriately; as you can see there is a test failure in your code, and the resolved conflicts lead to another bug in the master branch -  the function `_get_top_stocks` renamed to `get_top_stocks`; this called "software regression", see this for more information [regression-bug](https://diigo.com/0ipwxy)
in any case, don't merge until you figure out what's the problem.
I will resolve the bug, no problem

**Pull request merged**
